File: java/src/main/java/Queens.java
// Copyright 2017 John Hurst
// John Hurst (john.b.hurst@gmail.com)
// 2017-09-19

import java.time.Duration;
import java.time.Instant;

public class Queens {

  public static void main(String[] args) {
    if (args.length != 1 && args.length != 2) {
      usage();
    }
    int from = Integer.parseInt(args[0]);
    int to = args.length == 1 ? from : Integer.parseInt(args[1]);
    for (int size = from; size <= to; size++) {
      Instant start = Instant.now();
      Board board = new Board(size);
      int count = board.solve();
      Instant end = Instant.now();
      Duration duration = Duration.between(start, end);
      System.out.println("Board size " + size + " has " + count + " solutions. Calculated in " + duration + ".");
    }
  }

  private static void usage() {
    System.err.println("Usage: Queens fromSize [toSize]");
    System.exit(1);
  }
}

File: fsharp/Queens.fs
// Copyright 2018 John Hurst
// John Hurst (john.b.hurst@gmail.com)
// 2018-01-28

open System
open System.Diagnostics

type Board = {Size: int; Row: int; Cols: int; Diags1: int; Diags2: int;}

let empty size = {Size = size; Row = 0; Cols = 0; Diags1 = 0; Diags2 = 0}

let ok board col =
    ((board.Cols &&& (1 <<< col)) |||
     (board.Diags1 &&& (1 <<< board.Row + col)) |||
     (board.Diags2 &&& (1 <<< board.Row - col + board.Size - 1))) = 0

let place board col = {
    board with
        Row = board.Row + 1 ;
        Cols = board.Cols ||| (1 <<< col);
        Diags1 = board.Diags1 ||| (1 <<< board.Row + col);
        Diags2 = board.Diags2 ||| (1 <<< board.Row - col + board.Size - 1)
}

let rec solve board =
    if board.Row = board.Size then 1
        else seq { 0 .. board.Size - 1 }
            |> Seq.filter (ok board)
            |> Seq.map (place board)
            |> Seq.map solve
            |> Seq.sum

[<EntryPoint>]
let main argv =
    let first = if argv.Length = 0 then 8 else int argv.[0]
    let last = if argv.Length = 1 then first else int argv.[1]
    for size = first to last do
        let watch = Stopwatch.StartNew()
        let count = size |> empty |> solve
        watch.Stop()
        let ts = watch.Elapsed
        printfn "Board size %d has %d solutions.  Calculated in %02d:%02d:%02d.%03d." size count ts.Hours ts.Minutes ts.Seconds ts.Milliseconds
    0
